### PR TITLE
Add chapter about consolidation

### LINKED ISSUE #12: [Chapter] UTXO consolidation
https://en.bitcoin.it/wiki/Techniques_to_reduce_transaction_fees#Consolidation

https://bitcoinops.org/en/xapo-utxo-consolidation/

### PR DESCRIPTION
**Pull request by harding**
Closes #12 

Various transaction size calculations were performed using the tool in https://github.com/bitcoinops/bitcoinops.github.io/pull/322

### CODE REVIEW

File: x.consolidation/consolidation.md [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+![Savings by consolidation feerate](img/estimating-savings-by-feerate.png)
+
+For those that spend an average of two UTXOs or more per transaction
+and who can consolidate at 50% or less of their normal feerate (and who
+use the P2WPKH transactions analyzed here), consolidation at least
+breaks even on costs and can easily save more money than sending a
+normal transaction.  The feerate data included in the appendix to this
+chapter shows that saving at least 50% on feerates for a nominal 2-block
+confirmation target was possible 99.85% of the time during our analysis
+period of November and December 2019.
```
Comment (bitschmidty) [inline x.consolidation/consolidation.md]: How difficult is it to expand this analysis to years instead of months?

File: x.consolidation/consolidation.md [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
+Again, the predictive/organic method does have the advantage that it may
+be significantly easier to implement than consolidation for existing
+services, and combined with changeless transactions it may achieve a
+modest additional savings plus boost privacy.  Still, spenders focused
+on maximizing fee savings will probably want to focus on UTXO consolidation.
+
+## Fee bumping
+
+The maximize savings, consolidation transactions should be sent at the
```
Comment (bitschmidty) [inline x.consolidation/consolidation.md]: s/The/To

File: x.consolidation/consolidation.md [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+## See also
+
+- [Field report: consolidation of 4 million UTXOs at Xapo][] by Anthony
+  Towns
+
+- [Presentation: unspent management and coin selection][] by Mark
+  Erhardt
+
+## Appendix A:
+
```
Comment (bitschmidty) [inline x.consolidation/consolidation.md]: What about times of day to consolidate based on data?
or days of week to consolidate based on data?

File: x.consolidation/consolidation.md [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+  consolidation unless you were planning to transfer funds to your hot
+  wallet anyway.
+
+## Conclusion 
+
+For organizations that both frequently receive and frequently send
+payments, consolidation is one of the most efficient ways to save money.
+It's relatively easy to implement and, if done well, the people
+receiving payments from you will never notice that you're doing it.
+
```
Comment (bitschmidty) [inline x.consolidation/consolidation.md]: What about consolidating while spending (adding 'unnecessary' inputs to the current transaction which are consolidated into a larger change output). And even moving from hot to cold at the same time.

File: x.consolidation/consolidation.md [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+[avoidpartialspends]: https://bitcoinops.org/en/newsletters/2018/07/31/#bitcoin-core-12257
+[Presentation: unspent management and coin selection]: https://www.youtube.com/watch?v=hrlNN3BSB6w&feature=emb_logo
+[Field report: consolidation of 4 million UTXOs at xapo]: https://bitcoinops.org/en/xapo-utxo-consolidation/
+[payment batching]: https://github.com/bitcoinops/scaling-book/blob/master/x.payment_batching/payment_batching.md
+[channel factories]: http://devbcop/en/topics/channel-factories/
+[joinpools]: https://freenode.irclog.whitequark.org/bitcoin-wizards/2019-05-21#24639483;
+[BIP125]: https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki
+[predictive utxo management]: https://www.businesswire.com/news/home/20180726005483/en/BitGo-Introduces-Predictive-UTXO-Management-Delivering-30
+[organic utxo consolidation]: https://www.reddit.com/r/Bitcoin/comments/9kxn01/organic_utxo_consolidation_in_the_course_of/
```
Comment (bitschmidty) [inline x.consolidation/consolidation.md]: Potential alternate link? https://medium.com/simplexum/on-predictive-utxo-management-974be49506ca

File: x.consolidation/consolidation.md [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+Unspent Transaction Output (UTXO) consolidation is the act of taking the
+bitcoin value stored in several UTXOs and *consolidating* it into one
+UTXO.  This lowers the number of UTXOs you control while keeping your
+balance roughly the same, increasing the chance that you'll be able to
+create smaller and more affordable transactions at a later time when
+you need to pay a higher feerate in order to obtain a fast confirmation.
+
+## A consolidation example
+
+Imagine a transaction as a rectangle whose height is its feerate, whose
```
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: This is noted in a comment in the gnuplot file, but I deliberately avoided choosing a particular feerate here (and in the other plots, except the last one based on real data) because any feerate we use will probably not match the current feerates at the time people are reading this.  (I think we had this same problem in one of the other chapters, so I wonder if maybe it'd be an idea to invent bogocoins or bogobits or something, describe them in the introduction, and then use them throughout the book.)

File: x.consolidation/consolidation.md [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+length is its size (block weight), and whose area is its total fee
+(feerate times size).  For example, illustrated below is a simple
+transaction that spends from one P2WPKH UTXO (input) to two P2WPKH
+outputs---one output to a client and one as change back to the spender.
+
+![Example of a 1-input, 2-output transaction](img/consolidation-example-1in2out.png)
```
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: I don't understand this comment.  Are you asking for more information in the graphic or less?

File: x.consolidation/consolidation.md [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+example transaction above:
+
+![Example of a 3-input, 1-output transaction followed by a 1-input, 2-output transaction](img/consolidation-example-3in1out-1in2out.png)
+
+Comparing the total costs, we find that the two transaction version is
+40% cheaper than the single high-feerate transaction with three
+UTXOs.  This is possible even though the two-transaction version uses
+more total block space.
+
+If the consolidation transaction confirms before you send the second
```
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: > it seems somewhat implied that the consolidating tx would be confirmed

Yeah, it was implied earlier, but here I make that assumption explicit for accuracy.  It's also important because, later in the chapter, we describe fee bumping strategies to deal with consolidations that haven't confirmed yet.  I'll see if I can rephrase this.

File: x.consolidation/consolidation.md [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+For this reason, UTXO consolidation as described in this document should
+not be used by users trying to maximize the privacy of their bitcoin
+balance or transaction history.  It should especially be avoided by
+users who have previously used techniques such as coinjoin to conflate
+the ownership history of their UTXOs with other UTXOs.
+
+It may be possible to perform a limited amount of consolidation during a
+coinjoin, e.g. as implemented in Joinmarket or Wasabi.  Additionally,
+ideas for future protocol development such as [channel factories][] or
+[joinpools][] may make it more common for multiple UTXOs owned by
```
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: Yeah, the link I send people about it is the description we pulled last minute from our COSHV description last year: https://github.com/bitcoinops/bitcoinops.github.io/compare/1bb7eba1c611c6ac2b0155da47ce12b735f86099..8c3967226d0fdab46bc69bb993dcef6ef1da2e47#diff-682d5e69c5dbc0de911ac7e1213968f5L276

However, that feels inappropriate for a reference in a formal document, so I think the above link to Maxwell's original description is the best.  Maybe if I get bored some day, I'll write a description for the mailing list and send it to Maxwell for review; I am fond of the idea.

File: x.consolidation/consolidation.md (hunk 1 above)
Comment (harding) [reply x.consolidation/consolidation.md]: Trivially easy, but I can only collect data going forward, so I won't have a full year of data until mid-October 2020.

File: x.consolidation/consolidation.md (hunk 3 above)
Comment (harding) [reply x.consolidation/consolidation.md]: I made some charts for that two years ago: https://en.bitcoin.it/wiki/Miner_fees#The_market_for_block_space

I think we should save that kind of detailed information for Urgency Tradeoffs chapter, #15 .  I'll probably have a couple more months of logged fee estimates by that time too.

File: x.consolidation/consolidation.md (hunk 4 above)
Comment (harding) [reply x.consolidation/consolidation.md]: Consolidating while spending is the Predictive/Organic UTXO Management technique.  I'll see if I can make that clearer in that section.

I think it should be pretty rare that a high-frequency spender moves funds from hot to cold---for security, they should be receiving new deposits/payments to their cold wallet and then only moving enough money to the hot wallet to cover a limited number of outgoing payments/withdrawals (e.g. a week's worth).  I think you'd only want to move from hot to cold if there was emergency or maybe a huge price increase.

[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: I assume your premise is, e.g., Alice has 100k P2PKH UTXOs; when she consolidates, she can move that value to P2WPKH UTXOs.  Although that is a way to increase Alice's savings, I think from a safety engineering standpoint, we'd probably want to suggest that Alice make the address change and the consolidation change independently.  This is especially true given that the additional savings of doing the two together is pretty small:

- 14,800,000 vbytes --- 100,000 p2pkh UTXOs spend-to-selfs (inputs)
- 3,400 vbytes --- 100 p2pkh outputs
- 3,100 vbytes --- 100 p2wpkh outputs
- 14,800 vbytes --- 100 p2pkh UTXO payments (inputs)
- 6,775 vbytes --- 100 p2wpkh UTXO payments (inputs)

So, if we assume a consolidation feerate of 10 nBTC/vbyte and a payment feerate of 100 nBTC/vbyte,

- 0.0149514 BTC --- overall cost for not changing to segwit, `((14800000 + 3400) * 1 + 14800 * 10) / 1e9`
- 0.01487085 BTC --- overall cost if changed to segwit `((14800000 + 3100) * 1 + 6775 * 10) / 1e9`
- Segwit savings: 0.5%

Obviously, the amount of savings will vary depending on how many UTXOs Alice consolidates and what scripts she uses (e.g., she'll save more going from 2-of-3 P2SH TO 2-of-3 P2WSH), but I think that it's always going to be a small amount in percentage terms for consolidating a large number of UTXOs or a small amount in absolute value for consolidating a small number of UTXOs, so it isn't worth the risk of trying to make two changes to a money-handling system at the same time.

File: x.consolidation/consolidation.md [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+In short, if Alice can consolidate at a feerate of 1/4 of her regular
+spending rate or lower, she'll save more money than Bob even though she
+uses almost double as much block space as him.  The feerate data in the
+appendix to this chapter indicates such savings are possible under our
+rough assumptions over 96% of the time.
+
+If Bob needed to use more than two inputs per transaction in order to create
+a changeless transaction, Alice would be able to save more money even if she
+used a correspondingly-higher feerate for her consolidation transaction.
+
```
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: I don't understand this comment.  In this example, Bob is using his UTXOs for changeless payments.

File: x.consolidation/consolidation.md (hunk 3 above)
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Sounds good on punting this to another chapter

File: x.consolidation/consolidation.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: I agree things like feerate will not age well.

I dont have a strong preference on this so marking resolved.

File: x.consolidation/consolidation.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: This is related to the units discussion of having sat/vbyte. But Im ok leaving agnostic per other comment thread.

File: x.consolidation/consolidation.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Rewording looks good.

File: x.consolidation/consolidation.md [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+![Savings from consolidation](img/estimating-savings-p2wpkh-savings.png)
+
+Obviously the actual amount we save is going to depend on the ratio
+between the low feerate used for consolidations and the higher feerate
+used to pay clients.  Using data provided in the appendix of this
+chapter, the following ratios are commonly achievable as of this
+writing:
+
+![Savings by consolidation feerate](img/estimating-savings-by-feerate.png)
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Fixed

File: x.consolidation/consolidation.md [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+valid, that it pays an appropriate feerate, that it is less than 100,000
+vbytes, and that it follows a few other easy rules.  <!-- e.g., don't
+use patterns reserved for future soft forks -->
+
+An unconfirmed UTXO is the output of a transaction which is still in the
+mempool (i.e. has not yet been confirmed).  To prevent DoS attacks
+against node mempools, nodes place additional restrictions on
+transactions that use unconfirmed UTXOs.  That means you aren't
+guaranteed to be able to spend unconfirmed UTXOs at an arbitrary time.
+For example, a wallet may not be able to propagate the second
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Awesome, thanks for clarifying that.

File: x.consolidation/consolidation.md [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+   but they can save you some money.  When optimizing your UTXO pool for
+   changeless transactions, you want your UTXOs to have a variety of
+   amounts so you have as many different combinations of amounts as
+   possible.
+
+## Optimum consolidation size
+
+To consolidate UTXOs, you need to include more data than just the UTXOs
+in a transaction---you also have to include an output that pays the
+funds back to yourself plus some transaction boilerplate (version
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Fixed.

File: x.consolidation/consolidation.md [hunk 14]
```diff
@@ -0,0 +1,517 @@
...
+
+3. Focus on creating *changeless transactions* as described later in
+   this chapter.  For the reasons described in that later subsection,
+   changeless transactions usually aren't as efficient as consolidation,
+   but they can save you some money.  When optimizing your UTXO pool for
+   changeless transactions, you want your UTXOs to have a variety of
+   amounts so you have as many different combinations of amounts as
+   possible.
+
+## Optimum consolidation size
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Fixed

File: x.consolidation/consolidation.md [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+
+If you rarely encounter that problem with your UTXO pool, you can
+probably just ignore the few cases where it does happen as a minor
+inefficiency.  If it happens frequently, you have several options:
+
+1. Increase the average value of the outputs in your UTXO pool by
+   keeping the number of outputs the same and increasing the value of
+   each output, overall increasing the amount of money you have in the
+   pool.  Note that this also increases the amount of money you'll lose
+   due to theft or accident involving your hot wallet.
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved

File: x.consolidation/consolidation.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Fine by me.

File: x.consolidation/consolidation.md [hunk 16]
```diff
@@ -0,0 +1,517 @@
...
+consolidation transaction to confirm is easy for many organizations.
+Forthcoming chapters about feerate estimation and urgency tradeoffs will
+provide a more detailed analysis of the differences in feerates between
+different confirmation targets.
+
+[avoidpartialspends]: https://bitcoinops.org/en/newsletters/2018/07/31/#bitcoin-core-12257
+[Presentation: unspent management and coin selection]: https://www.youtube.com/watch?v=hrlNN3BSB6w&feature=emb_logo
+[Field report: consolidation of 4 million UTXOs at xapo]: https://bitcoinops.org/en/xapo-utxo-consolidation/
+[payment batching]: https://github.com/bitcoinops/scaling-book/blob/master/x.payment_batching/payment_batching.md
+[channel factories]: http://devbcop/en/topics/channel-factories/
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved

File: x.consolidation/consolidation.md [hunk 17]
```diff
@@ -0,0 +1,517 @@
...
+rough assumptions over 96% of the time.
+
+If Bob needed to use more than two inputs per transaction in order to create
+a changeless transaction, Alice would be able to save more money even if she
+used a correspondingly-higher feerate for her consolidation transaction.
+
+That said, a clear advantage of changeless transactions over
+consolidation transactions is that they allow the spender to save block
+space and achieve modest fee savings while preserving their privacy
+against the problems described in the *Consolidation Privacy* section
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved

File: x.consolidation/consolidation.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Im trying to articulate that the chance of Bob even being able to execute a changeless payment due to amount mismatches (and having to overpay in order to achieve a changeless payment?) is probably low in practicality.

File: x.consolidation/consolidation.md (hunk 1 above)
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Ill resolve this comment for now and we can revisit in the future when it makes sense.

File: x.consolidation/consolidation.md (hunk 4 above)
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Perhaps this miss was just my lack of understanding of the predictive/organic utxo mgmt techniques.

File: x.consolidation/consolidation.md (hunk 5 above)
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved.

File: x.consolidation/consolidation.md [hunk 18]
```diff
@@ -0,0 +1,517 @@
+# UTXO Consolidation
+
+Unspent Transaction Output (UTXO) consolidation is the act of taking the
+bitcoin value stored in several UTXOs and *consolidating* it into one
+UTXO.  This lowers the number of UTXOs you control while keeping your
+balance roughly the same, increasing the chance that you'll be able to
+create smaller and more affordable transactions at a later time when
+you need to pay a higher feerate in order to obtain a fast confirmation.
```
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved.

File: x.consolidation/consolidation.md (hunk 2 above)
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved.

File: x.consolidation/consolidation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: This was very informative, thanks @harding

Comment (harding) [reply x.consolidation/consolidation.md]: I did add a sentence to the first paragraph of the predictive/organic section to better describe how it works so hopefully other people won't encounter the same confusion.

Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Resolved

[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: Addressed here: https://github.com/bitcoinops/scaling-book/pull/24/commits/e0ed7e20fb5859ffa8bf6f494c416837c1af0620#diff-e9ba84d872ddd5e4388022fcd7ab047fR303

File: x.consolidation/consolidation.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: Addressed here (under the alternative name "payjoin"): https://github.com/bitcoinops/scaling-book/pull/24/commits/e0ed7e20fb5859ffa8bf6f494c416837c1af0620#diff-e9ba84d872ddd5e4388022fcd7ab047fR292

File: x.consolidation/consolidation.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (harding) [reply x.consolidation/consolidation.md]: In the provided example case where Bob only has 10 outputs to choose from (and they're all equal value), that absolutely is true.  In practice, a high-frequency spender who's also a high-frequency receiver will have a lot of UTXOs to choose from and those UTXOs may vary in value significantly, allowing them to take advantage of combinatorial blowup for finding good matches (or at least exponential increases if the spender is trying to limit their number of inputs to, e.g., 2).

For example, if Bob has `x` UTXOs each with a unique value, here's how many possible combinations there are of those UTXOs.  The more combinations, the more likely it is that Bob will find a combination that's close enough to his desired output value that he'll be able to produce a changeless transaction.



In short, I'm not sure we want to imply that creating changeless transactions is unlikely, at least when applied to high-frequency spenders.  (For regular spenders who only have a few UTXOs at most, yeah it's unlikely.)

[the remark this answers is not in the document]
Comment (bitschmidty) [reply x.consolidation/consolidation.md]: Yep we are on the same page re: your explanation.

It may be worth adding clarifying language around the example in your text vs the "practical" example outlined in your comment above.
